File: packages/server-php/src/MemoryCredentialStore.php
<?php

declare(strict_types=1);

namespace OpenPasskey\Server;

class MemoryCredentialStore implements CredentialStore
{
    /** @var StoredCredential[] */
    private array $credentials = [];

    public function store(StoredCredential $cred): void
    {
        $this->credentials[] = $cred;
    }

    public function get(string $credentialId): StoredCredential
    {
        foreach ($this->credentials as $cred) {
            if ($cred->credentialId === $credentialId) {
                return $cred;
            }
        }
        throw new PasskeyError('credential not found');
    }

    public function getByUser(string $userId): array
    {
        return array_values(array_filter(
            $this->credentials,
            fn(StoredCredential $c) => $c->userId === $userId,
        ));
    }

    public function update(StoredCredential $cred): void
    {
        foreach ($this->credentials as $i => $existing) {
            if ($existing->credentialId === $cred->credentialId) {
                $this->credentials[$i] = $cred;
                return;
            }
        }
        throw new PasskeyError('credential not found');
    }

    public function delete(string $credentialId): void
    {
        foreach ($this->credentials as $i => $cred) {
            if ($cred->credentialId === $credentialId) {
                array_splice($this->credentials, $i, 1);
                return;
            }
        }
        throw new PasskeyError('credential not found');
    }
}

File: packages/server-rails/lib/open_passkey/stores/memory_credential_store.rb
# frozen_string_literal: true

module OpenPasskey
  class MemoryCredentialStore
    include CredentialStore

    def initialize
      @credentials = []
    end

    def store(credential)
      @credentials << credential
    end

    def get(credential_id)
      cred = @credentials.find { |c| c.credential_id == credential_id }
      raise PasskeyError.new("credential not found") unless cred
      cred
    end

    def get_by_user(user_id)
      @credentials.select { |c| c.user_id == user_id }
    end

    def update(credential)
      idx = @credentials.index { |c| c.credential_id == credential.credential_id }
      raise PasskeyError.new("credential not found") unless idx
      @credentials[idx] = credential
    end

    def delete(credential_id)
      idx = @credentials.index { |c| c.credential_id == credential_id }
      raise PasskeyError.new("credential not found") unless idx
      @credentials.delete_at(idx)
    end
  end
end
